File: src/commenteme/collectors/power.py
"""Energia/frequência via `powermetrics` (EXIGE sudo).

powermetrics escreve amostras em formato plist no stdout, separadas por um byte
nulo (\\x00). Rodamos como subprocesso de longa duração e lemos o stream numa
thread, publicando sempre a última amostra parseada em self.latest.

Os campos de potência vêm em miliwatts; convertemos para watts.
"""

from __future__ import annotations

import plistlib
import subprocess
import threading


class PowerCollector:
    def __init__(self, interval_ms: int = 1000):
        self.interval_ms = interval_ms
        self.latest: dict | None = None
        self.error: str | None = None
        self._proc: subprocess.Popen | None = None
        self._thread: threading.Thread | None = None
        self._stop = threading.Event()

    def start(self) -> None:
        cmd = [
            "sudo",
            "powermetrics",
            "--samplers",
            "cpu_power,gpu_power,thermal",
            "-i",
            str(self.interval_ms),
            "-f",
            "plist",
        ]
        try:
            self._proc = subprocess.Popen(
                cmd,
                stdout=subprocess.PIPE,
                stderr=subprocess.DEVNULL,
            )
        except Exception as e:  # noqa: BLE001
            self.error = f"falha ao iniciar powermetrics: {e}"
            return
        self._thread = threading.Thread(target=self._reader, daemon=True)
        self._thread.start()

    def _reader(self) -> None:
        buf = b""
        stdout = self._proc.stdout
        try:
            for chunk in iter(lambda: stdout.read(4096), b""):
                if self._stop.is_set():
                    break
                buf += chunk
                while b"\x00" in buf:
                    raw, buf = buf.split(b"\x00", 1)
                    raw = raw.strip()
                    if not raw:
                        continue
                    try:
                        self.latest = plistlib.loads(raw)
                    except Exception:  # noqa: BLE001 — amostra parcial; ignora
                        pass
        except Exception as e:  # noqa: BLE001
            self.error = str(e)

    @staticmethod
    def _mw_to_w(d: dict, key: str):
        v = d.get(key)
        return (v / 1000.0) if isinstance(v, (int, float)) else None

    def read(self) -> dict | None:
        d = self.latest
        if not d:
            return None
        proc = d.get("processor", {}) or {}
        gpu = d.get("gpu", {}) or {}
        freq_hz = gpu.get("freq_hz")
        return {
            "cpu_w": self._mw_to_w(proc, "cpu_power"),
            "gpu_w": self._mw_to_w(proc, "gpu_power"),
            "ane_w": self._mw_to_w(proc, "ane_power"),
            "package_w": self._mw_to_w(proc, "combined_power")
            or self._mw_to_w(proc, "package_power"),
            "gpu_freq_mhz": (freq_hz / 1e6) if isinstance(freq_hz, (int, float)) else None,
            "thermal": d.get("thermal_pressure"),
        }

    def stop(self) -> None:
        self._stop.set()
        if self._proc:
            try:
                self._proc.terminate()
            except Exception:  # noqa: BLE001
                pass


if __name__ == "__main__":  # debug: sudo PYTHONPATH=src .venv/bin/python -m commenteme.collectors.power
    import time

    c = PowerCollector()
    c.start()
    print("aguardando powermetrics (precisa de sudo)…")
    for _ in range(8):
        time.sleep(0.5)
        r = c.read()
        if r:
            print(r)
            break
    else:
        print("sem amostra — rode com sudo? erro:", c.error)
    c.stop()

File: src/collectors/power.ts
/**
 * Power / frequency via `powermetrics` (REQUIRES sudo).
 *
 * powermetrics streams plist samples on stdout separated by a NUL byte (\x00).
 * We run it as a long-lived subprocess and keep the latest parsed sample. Power
 * fields arrive in milliwatts; we convert to watts.
 */

import { spawn, type ChildProcess } from "node:child_process";
import * as plist from "plist";
import bplist from "bplist-parser";
import type { PowerData } from "./types.js";

function parseSample(raw: Buffer): any | null {
  if (raw.length === 0) return null;
  try {
    // binary plist?
    if (raw.length >= 6 && raw.subarray(0, 6).toString("latin1") === "bplist") {
      const arr = bplist.parseBuffer(raw);
      return arr?.[0] ?? null;
    }
    // XML plist
    const text = raw.toString("utf8");
    const start = text.indexOf("<?xml");
    if (start === -1) return null;
    return plist.parse(text.slice(start));
  } catch {
    return null;
  }
}

function mwToW(d: any, key: string): number | null {
  const v = d?.[key];
  return typeof v === "number" ? v / 1000 : null;
}

export class PowerCollector {
  private latest: any = null;
  error: string | null = null;
  private proc: ChildProcess | null = null;
  private buf: Buffer = Buffer.alloc(0);
  private stopped = false;

  constructor(private intervalMs = 1000) {}

  start(): void {
    try {
      this.proc = spawn(
        "sudo",
        [
          "powermetrics",
          "--samplers",
          "cpu_power,gpu_power,thermal",
          "-i",
          String(this.intervalMs),
          "-f",
          "plist",
        ],
        { stdio: ["ignore", "pipe", "ignore"] },
      );
    } catch (e) {
      this.error = `failed to start powermetrics: ${e}`;
      return;
    }

    this.proc.stdout?.on("data", (chunk: Buffer) => {
      if (this.stopped) return;
      this.buf = Buffer.concat([this.buf, chunk]);
      let nul: number;
      while ((nul = this.buf.indexOf(0x00)) !== -1) {
        const raw = this.buf.subarray(0, nul);
        this.buf = this.buf.subarray(nul + 1);
        const sample = parseSample(raw);
        if (sample) this.latest = sample;
      }
    });

    this.proc.on("error", (e) => {
      this.error = String(e);
    });
  }

  read(): PowerData | null {
    const d = this.latest;
    if (!d) return null;
    const proc = d.processor ?? {};
    const gpu = d.gpu ?? {};
    const freqHz = gpu.freq_hz;
    return {
      cpuW: mwToW(proc, "cpu_power"),
      gpuW: mwToW(proc, "gpu_power"),
      aneW: mwToW(proc, "ane_power"),
      packageW: mwToW(proc, "combined_power") ?? mwToW(proc, "package_power"),
      gpuFreqMhz: typeof freqHz === "number" ? freqHz / 1e6 : null,
      thermal: d.thermal_pressure ?? null,
    };
  }

  stop(): void {
    this.stopped = true;
    if (this.proc) {
      try {
        this.proc.kill("SIGTERM");
      } catch {
        // ignore
      }
    }
  }
}
